HLRC: Add get users action

nknize opened the pull request:
This commit adds get users action to the high level rest client.

Relates #29827

elasticmachine commented:
Pinging @elastic/es-core-features

nknize commented:
Labeling this as WIP as its not entire complete but I wanted to get some immediate feedback to keep this task moving toward completion. I have one key question:

* The example in the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-user.html) have `"enabled" : true` in the response. But the user class didn't have an `enabled` field, only the `AuthenticateResponse` class did. I added `boolean enabled` to `Users` but I don't know if this is the right way to go. Furthermore, its causing a test failure now in `SecurityIT#testAuthenticate`

Other than that I'd like someone to have a quick look to verify I'm not missing anything. Once the above question is answered I think I just need to add an explicit test for calls to the `user` endpoint without any usernames explicitly provided.

nknize commented:
ping: @elastic/es-security

jaymode commented:
>  But the user class didn't have an enabled field, only the AuthenticateResponse class did. I added boolean enabled to Users but I don't know if this is the right way to go.

@albertzaharovits had originally put the enabled field on the user and then removed it from the object, see https://github.com/elastic/elasticsearch/pull/33552#issuecomment-432043192 for the reasoning.

jaymode commented:
I did not see anything jump out as missing to me

albertzaharovits commented on client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java:
```diff
@@ -75,6 +77,34 @@
+     * Get a user, or list of users, in the native realm synchronously.
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-user.html">
+     * the docs</a> for more information.
+     * @param request the request with the nuser's name
```
typo

nknize replied:
:+1:

albertzaharovits commented on client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java:
```diff
@@ -75,6 +77,34 @@
+     * Get a user, or list of users, in the native realm asynchronously.
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-user.html">
+     * the docs</a> for more information.
+     * @param request the request with the nuser's name
```
idem typo

nknize replied:
:+1:

albertzaharovits commented on client/rest-high-level/src/main/java/org/elasticsearch/client/security/GetUsersResponse.java:
```diff
@@ -0,0 +1,70 @@
+ * Returns a List of {@link User} objects
+ */
+public class GetUsersResponse {
+    private final List<User> users;
```
can we please make this a `Set` since the order is not important

nknize replied:
:+1:

albertzaharovits commented:
@nknize I would much prefer not to reintroduce `enabled` on the `User` on the client side. I don't believe this a view of a user that the client should be concerned about. It doesn't sound good to have two users that are otherwise equal but they have the `enabled` flag different. Are the users actually the same one or not? I would say they are, but something has changed on the server that makes the same user to behave differently at different time points; like changing his roles does not really alter the identity.

I think `GetUsers` should be returning two sets of users, the set of all users and the set of enabled users.

tvernum commented:
> I think GetUsers should be returning two sets of users, the set of all users and the set of enabled users.

I think that creates a strange API for clients that just want to iterate across all users (they'd have to loop twice, or we'd have to do some fancy merging of `Iterable`s).

I feel it would be a nicer API to just add a `isEnabled(User user)` (or `(String principal)`) method to the response object, and have a `Map` or `Set` backing that.

albertzaharovits commented:
> If we don't want the client to be concerned about enabled, then why would we return all users and enabled users? Wouldn't we just want to return either all users (if a username is not specified) or the requested user?

What I meant was that the current `enabled` status is not part of the user's identity. We still want the client to be able to check it, preferably without another round trip to the server so it should be exposed in the `GetUsersResponse`.

> I think that creates a strange API for clients that just want to iterate across all users (they'd have to loop twice, or we'd have to do some fancy merging of Iterables).

I was thinking of the following client experience:
```
for (User user : response.getUsers()) {
  if (response.getEnabledUsers().contains(user)) {
    // do smth with enabled user
  } else {
    // do smth with disabled user
  }
}
```
The enabled check as well be `response.isEnabled(User user)`, as Tim suggested, IMO both are about the same.

nknize commented:
I guess the main question I have here is: how do I know if a user is enabled or disabled if I do not add the `enabled` variable to the `User` class? It looks like only `AuthenticateResponse` and `PutUserRequest` have the enabled variable.

nknize commented:
@albertzaharovits @tvernum I removed the enabled flag from the User. I also moved the User parser to the `GetUserResponse` class and added an `enabledUsers` set. Is this closer to what y'all were thinking?

jkakavas commented on client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityRequestConverters.java:
```diff
@@ -67,6 +68,15 @@ static Request changePassword(ChangePasswordRequest changePasswordRequest) throw
 
+    static Request getUsers(GetUsersRequest getUsersRequest) {
+        RequestConverters.EndpointBuilder builder = new RequestConverters.EndpointBuilder()
+            .addPathPartAsIs("_xpack/security/user");
```
`/_xpack/security/user` - > `/_security/user` since https://github.com/elastic/elasticsearch/pull/36293 and https://github.com/elastic/elasticsearch/pull/36379 are merged

nknize replied:
:+1:

jkakavas commented on client/rest-high-level/src/test/java/org/elasticsearch/client/SecurityRequestConvertersTests.java:
```diff
@@ -101,6 +102,19 @@ public void testDeleteUser() {
+        final Map<String, String> expectedParams = getExpectedParamsFromRefreshPolicy(refreshPolicy);
+        Request request = SecurityRequestConverters.getUsers(getUsersRequest);
+        assertEquals(HttpGet.METHOD_NAME, request.getMethod());
+        assertEquals("/_xpack/security/user/test", request.getEndpoint());
```
`/_xpack/security/user/test` - > `/_security/user/test` since https://github.com/elastic/elasticsearch/pull/36293 and https://github.com/elastic/elasticsearch/pull/36379 are merged

nknize replied:
:+1:

albertzaharovits commented:
> I removed the enabled flag from the User. I also moved the User parser to the GetUserResponse class and added an enabledUsers set. Is this closer to what y'all were thinking?

Looks good, thank you!
I will take another look when you're ready for merge.

nknize commented:
retest this please

nknize commented:
This is probably okay to merge. The failures look unrelated to this change. Will keep trying to reach a passing CI result.

jkakavas commented:
```
00:08:42 Tests with failures:
00:08:42   - org.elasticsearch.client.SecurityRequestConvertersTests.testGetUsers
```
I left a suggestion inline for
```
00:08:40 FAILURE 0.02s J7 | SecurityRequestConvertersTests.testGetUsers <<< FAILURES!
00:08:40    > Throwable #1: org.junit.ComparisonFailure: expected:</[]_security/user/test> but was:</[/]_security/user/test>
```

The `org.elasticsearch.xpack.security.transport.netty4.SimpleSecurityNetty4ServerTransportTests.testHandshakeWithIncompatVersion` seems unrelated.

jkakavas commented on client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityRequestConverters.java:
```diff
@@ -67,6 +68,15 @@ static Request changePassword(ChangePasswordRequest changePasswordRequest) throw
 
+    static Request getUsers(GetUsersRequest getUsersRequest) {
+        RequestConverters.EndpointBuilder builder = new RequestConverters.EndpointBuilder()
+            .addPathPartAsIs("/_security/user");
```
```suggestion
            .addPathPartAsIs("_security/user");
```

albertzaharovits commented on client/rest-high-level/src/test/java/org/elasticsearch/client/SecurityIT.java:
```diff
@@ -89,6 +92,15 @@ public void testAuthenticate() throws Exception {
+        final GetUsersResponse getUsersResponse =
+            execute(getUsersRequest, securityClient::getUsers, securityClient::getUsersAsync);
+        ArrayList<User> usrs = new ArrayList<>();
+        usrs.addAll(getUsersResponse.getUsers());
```
typo

albertzaharovits commented on client/rest-high-level/src/test/java/org/elasticsearch/client/SecurityRequestConvertersTests.java:
```diff
@@ -101,6 +102,19 @@ public void testDeleteUser() {
     }
 
+    public void testGetUsers() {
+        final String[] users = new String[] {"test"};
```
I would test with a random array :)

nknize replied:
:+1:

albertzaharovits commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SecurityDocumentationIT.java:
```diff
@@ -109,6 +111,17 @@
+            assertNotNull(getUsersResponse.getUsers());
+        }
+    }
+
```
this is not complete yet, I would simply use `testGetRoles` as a blueprint.
The same with the documentation, use `./docs/java-rest/high-level/security/get-roles.asciidoc` as an example.
And don't forget to add it to `./x-pack/docs/en/rest-api/security.asciidoc`.

nknize replied:
:+1:

javanna commented:
This is labelled 6.6.0 but it has not been backported to 6.x. Just a reminder to backport #36622 once this is backported.
